gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/converter/EnvelopeSchemaConverter.java

```diff
@@ -17,136 +17,141 @@
 
 package org.apache.gobblin.converter;
 
-import com.google.common.base.Optional;
-import com.google.common.cache.CacheBuilder;
-import com.google.common.cache.CacheLoader;
-import com.google.common.cache.LoadingCache;
-import org.apache.gobblin.configuration.ConfigurationKeys;
-import org.apache.gobblin.configuration.WorkUnitState;
-import org.apache.gobblin.converter.filter.AvroProjectionConverter;
-import org.apache.gobblin.converter.filter.AvroSchemaFieldRemover;
-import org.apache.gobblin.metrics.kafka.KafkaSchemaRegistry;
-import org.apache.gobblin.metrics.kafka.KafkaSchemaRegistryFactory;
-import org.apache.gobblin.metrics.kafka.SchemaRegistryException;
-import org.apache.gobblin.util.AvroUtils;
-import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
-import java.util.concurrent.ExecutionException;
-import javax.xml.bind.DatatypeConverter;
+
 import org.apache.avro.Schema;
 import org.apache.avro.generic.GenericDatumReader;
 import org.apache.avro.generic.GenericRecord;
 import org.apache.avro.io.Decoder;
 import org.apache.avro.io.DecoderFactory;
+import org.apache.gobblin.configuration.WorkUnitState;
+import org.apache.gobblin.metrics.kafka.KafkaSchemaRegistry;
+import org.apache.gobblin.metrics.kafka.KafkaSchemaRegistryFactory;
+import org.apache.gobblin.util.AvroUtils;
+
+import com.google.common.base.Optional;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import javax.xml.bind.DatatypeConverter;
+
 
 /**
- * A converter for extracting schema/records from an envelope schema.
- * Input schema: envelope schema - must have fields payloadSchemaId (the schema registry key of the output
- *               schema) and payload (byte data for output record)
- * Input record: record corresponding to input schema
- * Output schema: schema obtained from schema registry using key provided in input record's {@link #PAYLOAD_SCHEMA_ID_FIELD}
- * Output record: record corresponding to output schema obtained from input record's {@link #PAYLOAD_FIELD} as bytes
+ * Base class for an envelope schema converter using {@link KafkaSchemaRegistry}
  */
-public class EnvelopeSchemaConverter extends Converter<Schema, String, GenericRecord, GenericRecord> {
```

Comment:
Why do we change the signature of this class? This seems to be backward incompatible if anyone is using this class before. What is the class that will represent the old behavior of existing EnvelopeSchemaConverter? It seems like EnvelopePayloadConverter != old EnvelopeSchemaConverter.

Maybe rename your EnvelopeSchemaConverter to BaseEnvelopeSchemaConverter to avoid name conflict?

Do we need a payload converter when upconvert is not needed? This is old EnvelopeSchemaConverter behavior, shall we keep it?

Reply:
Yes, it's a back incompatible change for 3 reasons
- The old converter returns `String` as the output schema, which makes it rarely used because the majority of our constructs use `Schema` as the schema
- The old converter doesn't have a clear name. It deserializes the payload and ignores the other fields in the envelope schema
- Upconverting to a certain schema is necessary if records of different schemas will be stored into one file, which would fail otherwise. It's like a standardization.